src/core/pos.ml

```diff
@@ -65,17 +40,32 @@ let make : popt -> 'a -> 'a loc =
 let in_pos : pos -> 'a -> 'a loc =
   fun p elt -> { elt ; pos = Some p }
 
+(** [end_pos po] creates a position from the end of position [po]. *)
+let end_pos : popt -> popt = fun po ->
+  match po with
+  | None -> None
+  | Some p ->
+    Some { p with start_line = p.end_line; start_col = p.end_line }
+
 (** [none elt] wraps [elt] in a ['a loc] structure without any specific source
     code position. *)
 let none : 'a -> 'a loc =
   fun elt -> { elt ; pos = None }
 
+(** [cat p1 p2] is the concatenation of positon [p1] and [p2]. If [p1] and
+    [p2] have different filenames, the resulting filename is [None]. *)
+let cat : pos -> pos -> pos = fun p1 p2 ->
+  { fname = if p1.fname <> p2.fname then None else p1.fname
```

Comment:
Instead of None, shouldn't we raise Invalid_argument instead?

Reply:
You're probably right indeed